### PR TITLE
[IMP]l10n_it_ddt : Extend the fatturapa xml invoice and make a link w…

### PR DESCRIPTION
**Pull request by anr-odoo**
DDT is a delivery document contains detailed info on the merchandise, the carrier and the delivery

Task: https://www.odoo.com/web#id=1914640&action=327&model=project.task&view_type=form&menu_id=4720

pad: https://pad.odoo.com/p/r.ac2f5121a2a262522240ac871b37b07c

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_it_edi/models/account_invoice.py [hunk 1]
```diff
@@ -171,6 +170,9 @@ def invoice_generate_xml(self):
             invoice.message_post(
                 body=(_("E-Invoice is generated on %s by %s") % (fields.Datetime.now(), self.env.user.display_name))
             )
+    def _render_template(self, template_values):
+        content = self.env.ref('l10n_it_edi.account_invoice_it_FatturaPA_export').render(template_values)
+        return content
```
Comment (jco-odoo) [inline addons/l10n_it_edi/models/account_invoice.py]: We should make a _get_export_values method and inherit that which recuperates the template_values instead

File: addons/l10n_it_edi/data/invoice_it_template.xml [hunk 2]
```diff
@@ -137,7 +137,7 @@
                                 <ImportoBollo t-esc="format_numbers(record.l10n_it_stamp_duty)"/>
                             </DatiBollo>
                         </DatiGeneraliDocumento>
-                        <DatiDDT t-if="record.l10n_it_ddt_id">
+                        <DatiDDT t-if="record.l10n_it_ddt_id" name="DDT">
```
Comment (jco-odoo) [inline addons/l10n_it_edi/data/invoice_it_template.xml]: Is this name necessary (problem in stable) as maybe there is already just one DatiDDT tag.

File: addons/l10n_it_ddt/data/l10n_it_ddt_template.xml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
+	 <template id="my_view_name" inherit_id="l10n_it_edi.account_invoice_it_FatturaPA_export">
+            <xpath expr='//DatiDDT[@name="DDT"]' position="after">
```
Comment (jco-odoo) [inline addons/l10n_it_ddt/data/l10n_it_ddt_template.xml]: Can we put a t-if if not ddt_id is defined on the invoice?

File: addons/l10n_it_ddt/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        line_count = 1
+        for line in self.invoice_line_ids.filtered(lambda l: not l.display_type):
+            for picking in line.picking_ids:
+                if picking not in ddt_dict:
+                    ddt_dict[picking] = []
+                ddt_dict[picking].append(line_count)
+            line_count += 1
+        for picking in ddt_dict:
+            if len(ddt_dict[picking]) == line_count-1:
+                ddt_dict[picking] = []
```
Comment (jco-odoo) [inline addons/l10n_it_ddt/models/account_invoice.py]: Because it has everything on it?  Maybe put a comment to explain.

**Review by jco-odoo — COMMENTED** [review]
This seems to be going in the right direction.  Needs to be merged with the other branch afterwards.

File: addons/l10n_it_ddt/data/l10n_it_ddt_template.xml (hunk 3 above)
Comment (anr-odoo) [reply addons/l10n_it_ddt/data/l10n_it_ddt_template.xml]: it done.

File: addons/l10n_it_edi/data/invoice_it_template.xml (hunk 2 above)
Comment (anr-odoo) [reply addons/l10n_it_edi/data/invoice_it_template.xml]: it done.

File: addons/l10n_it_edi/models/account_invoice.py (hunk 1 above)
Comment (anr-odoo) [reply addons/l10n_it_edi/models/account_invoice.py]: it done.

File: addons/l10n_it_ddt/models/account_invoice.py (hunk 4 above)
Comment (anr-odoo) [reply addons/l10n_it_ddt/models/account_invoice.py]: it done.

**Comment by eLBati** [comment]
@anr-odoo @jco-odoo Hi, please consider that a module named `l10n_it_ddt` already exists at https://github.com/OCA/l10n-italy/tree/12.0/l10n_it_ddt

Also, OCA Italian group is working an a refactoring for v13 at https://github.com/OCA/l10n-italy/issues/1499

Do you want to join efforts?

**Comment by jco-odoo** [comment]
Hi @eLBati , Good to hear from you!  Right now, in the standard modules, the l10n_it_edi module just has one field ddt_id and a separate object you can create for it.  (DDT no stock case)

In this module, we try to see the picking as the DDT itself (DDT semplificato I suppose) where we calculate through the Odoo standard fields which invoice is linked to which DDTs. 

We could call the module differently in v12 like l10n_it_stock_ddt or only put it as l10n_it_ddt in v13.  

It would be good to have the modules of OCA being able to re-use the Odoo standard rather than define everything again and we can certainly use some help in making the Italian localization better.

**Comment by jco-odoo** [comment]
@eLBati Please check my branch here https://github.com/odoo-dev/odoo/tree/master-l10n-it-ddt-anr-jco

**Comment by eLBati** [comment]
@jco-odoo requesting feedback from the OCA Italian group at https://github.com/OCA/l10n-italy/issues/1545

We'll keep you updated.

Thanks

**Comment by jco-odoo** [comment]
@eLBati  Things don't necessarily need to be very complicated.  Maybe we can put the format import/export of the invoice, ... in l10n_it and leave the mail system in l10n_it_edi, which might simplify dependencies.  Just an idea.

**Comment by eLBati** [comment]
@jco-odoo we are trying to evaluate your proposal

> It would be good to have the modules of OCA being able to re-use the Odoo standard rather than define everything again and we can certainly use some help in making the Italian localization better

that is a good idea but would have a huge impact in the OCA Italian modules, in terms of refactoring.

As you can see at https://github.com/OCA/l10n-italy/tree/12.0/ we have several e-invoicing modules (the ones containing `fatturapa` name, name used for historical reasons)

Odoo has `l10n_it_edi` module (born later) that is vary basic and does not cover all the use cases covered by the OCA modules.
Only analyzing the differences would require time.
Did you check the functionalities offered by the OCA modules? How many [XML nodes of e-invoice](https://www.agenziaentrate.gov.it/portale/documents/20143/288192/Rappresentazione+tabellare+fattura+ordinaria+3007.xls/7fcbf282-91a9-5b7e-abde-ec96a4b31f4c) are handled by `l10n_it_edi`?

Are you interested in having a complete e-invoicing solution, like the one offered by the OCA modules?

**Comment by jco-odoo** [comment]
@eLBati  We want to have the basics work fine and in a simple way for the user.  As usual with Odoo standard, we want to focus on usability rather than on supporting all of the features.  It is not the goal to get every feature and every node in the Fatturapa XML in Odoo standard.  Of course, if there are things that are easy to add and help a lot of people, we should do it. You can easily check the qweb in l10n_it_edi.  (which should be easy to extend too) Of course, there are people that will need more.

**Comment by sergiocorato** [comment]
Hi @jco-odoo , in OCA Italy maintainer group we are discussing the topic, and the feeling is that the modules implemented in core are useful only for a very limited use-cases (as you said, 'basics'). As a matter of fact, modules for electronic invoice were created and shared before law applicability from local maintainers, so all subject interested have installed them to work legally.
These modules of Odoo core can't substitute them for almost all the cases (perhaps single professional? no, missing withholding tax it seems...)
The policy of Odoo S.A. is now (I read) to [fully cover | to not cover at all] a function in Odoo CE, but these modules are an exception [to cover very partially], so we would appreciate a clarification: are you aware that these modules don't fulfill a 'real' use of e-invoice?
If so, there is a plan to cover a minumum real use? In this case, we could collaborate and work togheter to give a better solution to final user (now, the solution given is: 'uninstall l10n_it_edi wich is in conflict with OCA modules').
If not, we'll work as usual.

**Comment by jco-odoo** [comment]
@sergiocorato I would really like to have an open discussion about what we are missing to get more people going.  (we need more feedback)  Honestly, what we are afraid of is to have tons of specifications (reverse tax, riba, ...) arriving at us, at each time saying it is different in Italy than in the rest of the world (mostly, there is a country somewhere doing the same), and that in the end we can't do the most basic stuff.  

But e.g. if you say a lot of people need those withholding taxes, then we need to add them to the CoA and if there is special information needed for them to add in the XML, I don't see why we could not add it.  And this XML in l10n_it_edi is normally also easy to extend in any case. So, thanks for being open about this and the best way is to talk to each other.

**Comment by eLBati** [comment]
Hi @jco-odoo proceding one step at a time, we could start working on electronic invoicing and provide to you the requirements needed to be covered.

For this purpose, you could start checking the automated tests of the 2 main modules related to e-invoicing: `l10n_it_fatturapa_in` (reception) and `l10n_it_fatturapa_out` (generation).

Those tests cover the main use cases that must be handled by e-invoicing modules. You can obviously ask us for any clarification.

One step further: analyzing the dependencies of the above 2 modules and assuming to follow your more basic approach, I see some modules that maybe can be removed and excluded by your use cases:

- l10n_it_split_payment
- account_fiscal_year
- account_tax_balance
- date_range
- l10n_it_withholding_tax

**Comment by C3POdoo** [comment]
Dear @anr-odoo,

Thank you for your contribution but the version 12.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._

**Pull request closed without merging**